### PR TITLE
Core/Movement: Respect walk state when chasing

### LINKED ISSUE #20938: [3.3.5] Creatures always run when chasing
**Description:**

Creatures always use move type MOVE_RUN when MoveChase() is issued. 

**Current behaviour:**

Creatures always use MOVE_RUN when MoveChase() is issued. regardless of whether they are set to walk or not, which is the default for any attacking creature. 

**Expected behaviour:**

Not all creatures should run when chasing their targets. Thaladred the Darkener (http://www.wowhead.com/npc=20064/thaladred-the-darkener) for example, should walk while chasing players, but currently there is no way to achieve that, because https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp#L220 always disables walk mode.

**Steps to reproduce the problem:**

1. Make Thaladred the Darkener walk instead running - SetWalk(true)
2. Start Kael'thas encounter
3. Watch as Thaladred still runs after his targets at full speed

**Branch(es):**

3.3.5/likely both

https://github.com/TrinityCore/TrinityCore/commit/a422dad8702c779b5665891d9e651aef575f33f7

### PR DESCRIPTION
**Pull request by Sorikoff**
**Changes proposed:**
* Chase Movement Generator is modified to not to set walk mode to false on initialization.
* Chase Movement Generator is also modified to take into account current walk mode during each update tick, therefore allowing to chase in walking mode only when needed and set to do so.

**Target branch(es):** 3.3.5

**Issues addressed:** Closes #20938

**Tests performed:** Does build, works-in-game

### CODE REVIEW

**Comment by Shauren** [comment]
What about all existing scripts that modify walk/run mode? I would like you be sure you thought about them (and adjusted if needed, for example by removing walk flag if it was set for RP intro or something similar) when making a change to generic system like this and not pretended they don't exist.

**Comment by Sorikoff** [comment]
Fair enough. Adjusted walk mode for creatures using it. Needs some additional testing.

File: src/server/scripts/Northrend/FrozenHalls/HallsOfReflection/halls_of_reflection.cpp [hunk 1]
```diff
@@ -706,6 +707,7 @@ class npc_jaina_or_sylvanas_intro_hor : public CreatureScript
                             falric->AI()->Talk(SAY_FALRIC_INTRO_1);
                             falric->SetWalk(true);
                             falric->GetMotionMaster()->MovePoint(0, FalricPosition[1]);
+                            marwyn->SetWalk(false);
```
Comment (illfated) [inline src/server/scripts/Northrend/FrozenHalls/HallsOfReflection/halls_of_reflection.cpp]: Copy/paste error, replace `marwyn` with `falric` :-P

**Comment by Sorikoff** [comment]
Tested. Everything works now.

File: src/server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp [hunk 2]
```diff
@@ -1235,6 +1235,7 @@ class npc_tirion_fordring_tft : public CreatureScript
                     me->RemoveFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_GOSSIP);
                     me->SetWalk(true);
                     me->GetMotionMaster()->MovePoint(POINT_TIRION_INTRO, TirionIntro);
+                    me->SetWalk(false);
```
Comment (ccrs) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp]: uh? this should be called when chase movement starts, right?

**Review by ccrs — CHANGES_REQUESTED** [review]
This actually needs to check all `SetWalk` uses in scripts and even core.

`SetWalk(false)` should be used just before `MoveChase` call.
And if there isnt, analyze where/when the unit should actually start running.

File: src/server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp (hunk 2 above)
Comment (Shauren) [reply src/server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp]: tirion never chases anything

Comment (ccrs) [reply src/server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp]: edited the comment

**Comment by jackpoz** [comment]
I wonder if "Running when chasing" should be the default behavior and then some sort of flag changes the behavior to "Respect setWalk() when chasing too"

**Comment by robinsch** [comment]
It's distance based. Same as the "more speed the more distance you have" when following. Most video sources I found are about < 3 yards the pet walks or chases in walking mode.

**Comment by Sorikoff** [comment]
@jackpoz, I like your approach. However, I am unsure whether it would be better to keep flag in DB or it would be just fine to make additional method for C++ scripting? It looks to me that first option could be better since it would make adjusting of this a way more easier.

**Comment by jackpoz** [comment]
db flag usually sounds better for these kind of things. 
do you have a list of units where it would be already needed ? just to get an idea if the db solution will fit them all.
anyway it should be possible to add support to C++ too later on if needed (or maybe it's already supported to change db flags at runtime from C++ scripts)

**Comment by Sorikoff** [comment]
I know a couple of them:
1. [Dark Rider of Acherus](https://www.wowhead.com/npc=28768/dark-rider-of-acherus) (walks in scripted event)
3. [Thaladred the Darkener](https://www.wowhead.com/npc=20064/thaladred-the-darkener) (walks always)
2. [Lava Scion](https://www.wowhead.com/npc=53231/lava-scion) (does not belong to 3.3.5 branch)

**Comment by Killyana** [comment]
I asked for a similar future to be able to choose between walk/run for random_movements https://github.com/TrinityCore/TrinityCore/issues/19909 maybe the gist could help.

**Pull request closed without merging**